Unified NHP FreeSurferPipeline

TakJim opened the pull request:
This pull request adds the current NHP version of the FreeSurfer processing pipeline scripts.

Files Added
- Examples/Scripts/FreeSurferPipelineBatchNHP.sh – Batch execution script for NHP pipeline
- FreeSurfer/FreeSurferPipelineNHP.sh – Main entry point of the NHP FreeSurfer pipeline
- FreeSurfer/custom/SetUpFSNHP.sh – Environment setup script for NHP FreeSurfer processing
- FreeSurfer/custom/conf2hiresNHP
- FreeSurfer/custom/recon-all.v6.hiresNHP
- FreeSurfer/scripts/FreeSurferHiresPialNHP.sh
- FreeSurfer/scripts/FreeSurferHiresWhiteNHP.sh

TakJim committed:
Add FreeSurferPipelineNHP.sh and related NHP version scripts required for execution

takuya-hayashi commented:
Followings were used for FS5.3. Do we need them in this release...? 
 
FreeSurfer/scripts/FreeSurferHiresPialNHP.sh
FreeSurfer/scripts/FreeSurferHiresWhiteNHP.sh

glasserm commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1219 @@
+#!/bin/bash
```
One question would be how severe is the diff from the human pipeline for the base pipeline script, as this looks to be based on a relatively recent version of the human pipeline script?  Another is whether longitudinal mode is supported for NHP (or if it should be supported)?

glasserm commented:
> > Followings were used for FS5.3. Do we need them in this release...?
> > FreeSurfer/scripts/FreeSurferHiresPialNHP.sh FreeSurfer/scripts/FreeSurferHiresWhiteNHP.sh
> 
> Do you want to support FS5.3 or just a later version?

Will not support FS5.3, will be removed.

TakJim committed:
remove files for FS5.3

TakJim committed:
Modified to allow the user to specify the path

coalsont commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,101 @@
+echo FREESURFER_HOME=$FREESURFER_HOME
+echo PERL
+which perl
+perl -V | tail -11
```
It doesn't look like any of these scripts use perl.  Why is this here?

coalsont replied:
In fact, I don't think the Batch script really needs to report anything about its environment.  What matters is the environment the pipeline sees when it is launched.

coalsont commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,101 @@
+      --t2="$T2wImage" \
+      "$FLAIRopt" \
+      --species="$SPECIES" \
+      --runmode="$RunMode"" 
```
Quotes don't nest without other syntax in between.
```suggestion
  echo set -- --subject="$Subject" \
      --subjectDIR="$SubjectDIR" \
      --t1="$T1wImage" \
      --t1brain="$T1wImageBrain" \
      --t2="$T2wImage" \
      "$FLAIRopt" \
      --species="$SPECIES" \
      --runmode="$RunMode"
```

coalsont commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,101 @@
+      --t1="$T1wImage" \
+      --t1brain="$T1wImageBrain" \
+      --t2="$T2wImage" \
+      "$FLAIRopt" \
```
"" will be detected as an argument and cause an error.  Let's use the newer syntax instead:
```suggestion
  isFLAIR=false
  if [ -e "${StudyFolder}/${Subject}/T1w/T2w_acpc_dc_restore.nii.gz" ] ; then
      T2wImage="${StudyFolder}/${Subject}/T1w/T2w_acpc_dc_restore.nii.gz" #T2w FreeSurfer Input (Full Resolution)
      T2wType="${T2wType:=T2w}" # T2w, FLAIR. Default is T2w
      if [ "$T2wType" = "FLAIR" ] ; then 
         isFLAIR=true
      fi
  else
  	T2wImage="NONE"
	T2wType=NONE
  fi
  
     ${HCPPIPEDIR}/FreeSurfer/FreeSurferPipelineNHP.sh \
      --subject="$Subject" \
      --subjectDIR="$SubjectDIR" \
      --t1="$T1wImage" \
      --t1brain="$T1wImageBrain" \
      --t2="$T2wImage" \
      --flair="$isFLAIR" \
```

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1219 @@
+	exit 0
+fi
+
+${HCPPIPEDIR}/show_version
```
This is all cruft from the "main function" style of old `opts.shlib`, the start of the script uses newopts instead.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1219 @@
+
+# Species-specific environments imported from SPECIES4FS.cnf
+export SPECIES
+source ${HCPPIPEDIR}/FreeSurfer/custom/SetUpFSNHP.sh
```
Let's make this simpler.  Also, do *not* use `newopts.shlib` in the `SetUp...` this time (its already sourced and populated with arguments since this is a pipeline, not a batch file, and I am not going to support the potential collisions for that case...but also, `newopts.shlib` is not designed for things that are sourced in general, and is overcomplicated for a small number of arguments).
```suggestion
source "$HCPPIPEDIR"/FreeSurfer/custom/SetUpFSNHP.sh "$Species"
```

TakJim replied:
Fixed it.
819e0ffeb6a4d71d2b47cd98333e96ff5dd14f2b

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1219 @@
+    T2wImage=""
+fi
+
+# 既存のパラメータ処理部分の後に追加
```
Google says this means "Add after existing parameter processing part", so this comment should probably be removed.

Philosophically, I don't see an issue with having non-english comments, as long as an english translation of them is also provided.

TakJim replied:
I removed it. 2b23b8b1e75342c7b58b33e47ddee4acad25e771

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1219 @@
+fi
+log_Msg "num_cores: ${num_cores}"
+
+if [ "${existing_subject}" != "TRUE" ] ; then
```
With the current `opts_StringToBool` sanitizing and the longitudinal naming, this should be:
```suggestion
if ((! existing_session)) ; then
```

TakJim replied:
I fixed it. 5bd66b17068175d90975a30eb7d1fe0aa204ffc8

coalsont commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,268 @@
+  VariableSigma="3"
+  PialSigma="2"
+  SmoothNiter="1"
+  if [ "${p_flair}" != "TRUE" ] ; then                        # control T2 pial
```
Don't use legacy `opts.shlib` parsing details, pass in what you need to know as an argument.
```suggestion
  if ((isFLAIR)) ; then                        # control T2 pial
```
This particular conditional currently does the same thing on both branches, though.

TakJim replied:
Sorry, I have applied the same fix to Marmoset and the other modules as well.
916f3e778bddaa87f1817a224911a76c89105c81

TakJim committed:
remove environment report

TakJim committed:
remove unnecessary quotes

TakJim committed:
remove FLAIRopt and add isFLAIR.

TakJim committed:
fix indentation

TakJim commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,101 @@
+  fi
+  
+     ${HCPPIPEDIR}/FreeSurfer/FreeSurferPipelineNHP.sh \
+      --subject="$Subject" \
```
fixed it. 1182ff296cd613371b79ea24b9650a217f285923

TakJim committed:
remove old opts.shlib style

TakJim committed:
remoe old style parameter setting statement

TakJim committed:
fixed old style SPECIES setting statement

TakJim committed:
remote old style logging

TakJim committed:
remove unused parameter

TakJim committed:
always remove "$SubjectID".expert.opts

TakJim committed:
remove unnecessary comment

TakJim committed:
fix conditional statement

TakJim committed:
add parameter setting section

TakJim commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,268 @@
+# All rights reserved.
+
+SkullStripMethod=PreFS                                             # PreFS or FS. PreFS method is recommended for NHP - TH Feb 2024
+
```
I fixed it. c54bd0d2c38c72a2f7c3e9377a7d2efe19b485d4

TakJim committed:
not to use regular expression

TakJim committed:
fixed old style parameters

glasserm commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1140 @@
+# Show usage information
+opts_AddMandatory '--session' 'SessionID' 'session' "Session ID (required).  Used with --path input to create full path to root directory for all outputs generated as path/session" "--subject"
+
+opts_AddOptional '--session-dir' 'SessionDIR' 'session' 'path to session directory required, unless --existing-session is set' "--subject-dir"
```
Change back to subject not session.

TakJim committed:
fixed old style parameters

TakJim committed:
Add FLAIR argument to SetUpFSNHP.sh call

TakJim committed:
Revert variable name from “session” back to “subject”

TakJim committed:
Allow users to modify variables by editing Batch.sh

glasserm commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,156 @@
+
+StudyFolder="${HOME}/projects/Pipelines_ExampleData"
+Subjlist="100307"
+SPECIES="Human"
```
As above, let's have this set up for a macaque, not a human and not a human subject ID.

TakJim replied:
I made the same fix as in` PreFreeSurferPipelineBatchNHP.sh`.
63cde16c4a2fb02994048abbb8ffe833f9ca5016

coalsont commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,156 @@
+Usage () {
+    echo "$(basename $0) --StudyFolder=<path> --Subject=<id> --Species=<species> --RunMode=<mode> [--T2wType=<type>] [--EnvironmentScript=<path>]"
+    echo ""
+    echo "Required Options:"
```
I think all of these options are now optional, so we don't need a distinction between required vs not.

```suggestion
    echo "$(basename $0) --StudyFolder=<path> --Subject=<id> --Species=<species> --RunMode=<mode> [--T2wType=<type>] --EnvironmentScript=<path>"
    echo ""
    echo "Options:"
```

coalsont commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,156 @@
+if [ -z "$StudyFolder" ] || [ -z "$Subjlist" ] || [ -z "$T2wType" ]  || [ -z "$SPECIES" ] || [ -z "$RunMode" ]; then
+    echo "ERROR: Missing required parameters"
+    Usage
+fi
```
Since we expect most users to edit these in the batch script, I would call them "settings" instead of "parameters".  Most other batch scripts don't even bother checking this (or even have the ability to show a usage statement), but it doesn't hurt.
```suggestion
# Check required settings
if [ -z "$StudyFolder" ] || [ -z "$Subjlist" ] || [ -z "$T2wType" ]  || [ -z "$SPECIES" ] || [ -z "$RunMode" ]; then
    echo "ERROR: Missing required settings"
    Usage
fi
```

TakJim committed:
remove unnecessary compatibility flag

TakJim committed:
Update FreeSurferPipelineBatchNHP.sh to change subject list and species to NHP

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1140 @@
+# Show usage information
+opts_AddMandatory '--subject' 'SubjectID' 'subject' "Subject ID (required).  Used with --path input to create full path to root directory for all outputs generated as path/subject"
+
+opts_AddOptional '--subject-dir' 'SubjectDIR' 'subject' 'path to subject directory required, unless --existing-subject is set' "--subject-dir"
```
I removed it.
ee5b8f7ec5db7956dc04e1ff0033d7704a062327

TakJim committed:
Fix comment

TakJim committed:
Fix usage and add --EnvironmentScript to get_batch_options

TakJim committed:
Fix comment and error message

TakJim committed:
Remove unnecessary empty-string fallback for EnvironmentScript

coalsont commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,270 @@
+  GreySigmaFS5="5"                                            # FS5.3.0 HighResPial
+  GCAdir="${HCPPIPEDIR_Templates}/ChimpYerkes29"
+  GCA="RB_all_2008-03-26.gca"
+  AvgCurvTif="average.curvature.filled.buckner40.tif"
```
These will need to be in the repository for users to be able to run this.

takuya-hayashi replied:
@coalsont @glasserm 
I'm organising the global/template/NHP_NNP subdirectory (including Chimp) and will share it with you directly very soon. I've realised that I don't have Chimp's GCA and tif files. Do you have it on your side or did you used human's?

coalsont commented on FreeSurfer/FreeSurferPipeline.sh:
```diff
@@ -127,9 +127,9 @@ fi
+opts_AddMandatory '--session' 'SessionID' 'session' "Session ID (required).  Used with --session-dir input to create full path to root directory for all outputs generated as path/session" "--subject"
 
-opts_AddOptional '--session-dir' 'SessionDIR' 'session' 'path to session directory required, unless --existing-session is set' "--subject-dir"
+opts_AddMandatory '--session-dir' 'SessionDIR' 'session' 'path to session directory required' "--subjectDIR"
```
Please set your branch to 5ef54e77c432a71d2ad1000601fa739ebd976c2c, which is the rebase before you added a merge commit after it, and force-push it as-is (without pull or merge).  Please don't make any other commits until after this has been done.

TakJim replied:
Thanks for your instruction. I've set the branch to https://github.com/Washington-University/HCPpipelines/commit/5ef54e77c432a71d2ad1000601fa739ebd976c2c

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1140 @@
+opts_AddOptional '--processing-mode' 'ProcessingMode' 'HCPStyleData or LegacyStyleData' "Controls whether the HCP acquisition and processing guidelines should be treated as requirements.  'HCPStyleData' (the default) follows the processing steps described in Glasser et al. (2013) and requires 'HCP-Style' data acquistion.  'LegacyStyleData' allows additional processing functionality and use of som ...
+
+# NHP options
+opts_AddMandatory '--species' 'Species' 'Human|Chimp|MacaqueCyno|MacaqueRhes|MacaqueFusc|NightMonkey|Marmoset' "Species type (required).  Controls species-specific processing parameters" ""
```
Mandatory parameters don't take a default (optional parameters default to empty string, so you wouldn't need to specify `""` anyway, unless you were adding a legacy flag).
```suggestion
opts_AddMandatory '--species' 'Species' 'Human|Chimp|MacaqueCyno|MacaqueRhes|MacaqueFusc|NightMonkey|Marmoset' "Species type (required).  Controls species-specific processing parameters"
```

coalsont commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,270 @@
+    GCAdir="${HCPPIPEDIR_Templates}/NHP_NNP/${BrainTemplate}/fsaverage"
+    TemplateWMSkeleton="${HCPPIPEDIR_Templates}/NHP_NNP/${BrainTemplate}/MNINonLinear/${BrainTemplate}_wmskeleton_0.5mm.nii.gz"
+
+  elif [[ "$SPECIES" == *Rhesus* ]] ; then                      # Rhesus (Macaca mulatta)
```
The NHP pipeline help says that `MacaqueRhes` is a valid setting, but it will not test true here.

TakJim committed:
Update help info for species type in FreeSurferPipelineBatchNHP.sh

TakJim commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,156 @@
+    echo "  --StudyFolder: Path to the study folder containing subject data"
+    echo "  --Subject: Subject identifier (multiple subjects can be separated by space or @)"
+    echo "  --T2wType: T2w image type (T2w or FLAIR, default: T2w)"
+    echo "  --Species: Species type (Human, Chimp, MacaqueCyno, MacaqueRhes, MacaqueFusc, NightMonkey, Marmoset)"
```
We changed the names to MacaqueResus and MacaqueSnow to make them consistent with PreFreeSurfer.
313d1694c265ccd8c7f25aa8ec038f260ece75c0

TakJim committed:
remove unnecessary section

TakJim committed:
Update species options in FreeSurferPipelineNHP.sh

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+fi
+if [[ "$T1wImageBrain" == "NONE" ]]; then
+    T1wImage=""
+fi
```
Is this what was intended?
```suggestion
if [[ "$T1wImageBrain" == "NONE" ]]; then
    T1wImageBrain=""
fi
```

glasserm commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,156 @@
+        T2wType=NONE
+    fi
+  
+    ${HCPPIPEDIR}/FreeSurfer/FreeSurferPipelineNHP.sh \
```
Missing SGE submission.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+extra_reconall_args=(${extra_reconall_args_manual[@]+"${extra_reconall_args_manual[@]}"})
+
+#parse booleans
+flair=$(opts_ToBool "$flair")
```
```suggestion
flair=$(opts_StringToBool "$flair")
```

TakJim replied:
Fixed in c75350b9ff655c869bcb370923bb1cc5d95121df.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+	log_Msg "conf2hires expert opts: $c2hxopts"
+	echo "conf2hiresNHP $c2hxopts" >> "$SubjectDIR"/"$SubjectID".expert.opts
+fi
+ExpertOpts="-expert "$SubjectDIR"/"$SubjectID".expert.opts -xopts-overwrite"
```
`"a"$b"c"` doesn't nest like that, the variables end up unquoted.  Including in extra args array instead.
```suggestion
extra_reconall_args+=(-expert "$SubjectDIR"/"$SubjectID".expert.opts -xopts-overwrite)
```

TakJim replied:
Fixed in 5a11fd49b433e992364da807aed63040cde1c051.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+	recon_all_initrun=" -motioncor"
+	recon_all_initrun+="$recon_all_T1input"
+	recon_all_initrun+="$recon_all_T1braininput"
+	recon_all_initrun+="$recon_all_T2input"
```
Simplify and use array.
```suggestion
	recon_all_initrun=(-motioncor "$recon_all_T1input" "$recon_all_T1braininput" "$recon_all_T2input")
```

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+	recon_all_initrun+="$recon_all_T2input"
+
+	log_Msg "...recon_all_cmd: ${recon_all_cmd} ${recon_all_initrun} ${extra_reconall_args} ${ExpertOpts}"
+	${recon_all_cmd} ${recon_all_initrun} ${extra_reconall_args} ${ExpertOpts}
```
Whitespace-safe syntax for the command/argument arrays.  In this case, we know that none of the arrays are empty, so we don't need `set -eu` compatibility syntax for bash 4 and earlier.
```suggestion
	log_Msg "...recon_all_cmd: ${recon_all_cmd[*]} ${recon_all_initrun[*]} ${extra_reconall_args[*]}"
	"${recon_all_cmd[@]}" "${recon_all_initrun[@]}" "${extra_reconall_args[@]}"
```

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+		recon_all_T2input=""
+		recon_all_pial=""
+		T2Type="NONE"		
+fi
```
Array logic on the multiply-used, optional and flag-changing T2 argument.
```suggestion
if [ "${T2wImage}" != "NONE" ] ; then
	if [ "${flair}" = "TRUE" ] ; then
		recon_all_pial="-FLAIRpial"
		recon_all_T2input=(-FLAIR "$(remove_ext "$T2wImage")_scaled.nii.gz")
		T2Type=FLAIR
	else
		recon_all_pial="-T2pial"
		recon_all_T2input=(-T2 "$(remove_ext "$T2wImage")_scaled.nii.gz")
		T2Type=T2
	fi
	rm -f "$SubjectDIR"/"$SubjectID"/mri/transforms/"$T2Type"raw.lta # remove this otherwise conf2hires will not update this - TH
else
		recon_all_T2input=()
		recon_all_pial=""
		T2Type="NONE"		
fi
```

TakJim replied:
Fixed it in ee573dad64a8ef26545c0f1ddec4ae6caaaf4292.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+		${HCPPIPEDIR}/global/scripts/ScaleVolume.sh "$T1wImageBrain" "$ScaleFactor" $(remove_ext ${T1wImageBrain})_scaled 
+
+		recon_all_T1input=" -i $(remove_ext ${T1wImage})_scaled.nii.gz"
+		recon_all_T1braininput=" -emregmask $(remove_ext ${T1wImageBrain})_scaled.nii.gz"
```
These variables are no longer needed.
```suggestion
```

TakJim replied:
Fixed it in 458210c2dfaf02a955f4dc9257391eeafb83c9f7.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+# By default, refine pial surfaces using T2 (if T2w image provided).
+# If for some other reason the -T2pial flag needs to be excluded from recon-all, 
+# this can be accomplished using --extra-reconall-arg=-noT2pial
+if [ "${T2wImage}" != "NONE" ] ; then
```
"NONE" was converted to "" above.
```suggestion
if [[ "${T2wImage}" != "" ]] ; then
```

TakJim replied:
Fixed it in 1963bc8076f9c487cbaf10766764168defd8e1f1.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+		recon_all_T1input=" -i $(remove_ext ${T1wImage})_scaled.nii.gz"
+		recon_all_T1braininput=" -emregmask $(remove_ext ${T1wImageBrain})_scaled.nii.gz"
+
+		if [ "${T2wImage}" != "NONE" ] ; then
```
```suggestion
		if [[ "${T2wImage}" != "" ]] ; then
```

TakJim committed:
Add error handling for unmatched Macaque subspecies in SetUpFSNHP.sh

TakJim commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,270 @@
+    GCAdir="${HCPPIPEDIR_Templates}/NHP_NNP/${BrainTemplate}/fsaverage"
+    TemplateWMSkeleton="NONE"                                 # wmskeleton is not effective for snow monkey 
+
+  fi
```
Fixed in bf6f19a06db0923165e1854c04ca76799213f1a5.

coalsont replied:
That commit doesn't appear to be on this PR branch.  You appear to have committed it to a "NHP_FreeSurfer" branch, when the PR uses "NHP_FreeSurferPipeline".

coalsont replied:
I also notice that commit uses `=~` that we already commented on and fixed to be more obvious with `== *pattern*`.

coalsont replied:
It looks like you picked only one commit, rather than the range of 5 commits you made on the wrong branch (`NHP_FreeSurfer`).  Here are the 4 commits that did not get cherry-picked, some (but not all) of which you seem to have re-edited manually onto the correct branch:

https://github.com/RIKEN-BCIL/HCPpipelinesNHPmerge/compare/bf6f19a06db0923165e1854c04ca76799213f1a5...NHP_FreeSurfer

The `$SubjectID` should be `$Subject`, and you manually fixed the `T1wImageBrain` bug again, but the rest of those changes should be applied to the PR branch.

TakJim committed:
Correct T1wImageBrain assignment when T1wImageBrain is set to NONE

TakJim committed:
remove wrong flag

TakJim commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,156 @@
+        --subjectDIR="$SubjectDIR" \
+        --t1="$T1wImage" \
+        --t1brain="$T1wImageBrain" \
+        --t2="$T2wImage" \
```
Fixed in 1f540bebf071640685f48e3ee809508098461228 and 506e465ae6ceb2f4cd65f3c59214dc8e8542d4a2.

TakJim committed:
Fix: parse booleans

TakJim committed:
add variable initialization

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+# options for conf2hires
+if [ ! -z "$T1normSigma" ] ; then
+	c2hxopts=" --t1norm-sigma $T1normSigma"
+fi
```
Fixed it in 32013866c6e3d579268010bb2cad77dc86cc5010.

TakJim committed:
Fix the NHP recon-all calling

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+# to have the "final say" over a couple settings within recon-all
+if [ "${conf2hires}" = "TRUE" ] ; then
+	conf2hiresflag=" -conf2hires"
+fi
```
Thanks for your suggestion.
Fixed it in 
fd7725082aa75bb299ad8c90c87bfed4f687d64f.

TakJim committed:
Add extra syntax to tolerate unset variables

TakJim committed:
Fix the quote handling

TakJim committed:
Use arryay for recon_all_initrun

TakJim committed:
Fix for whitespace-safe syntax

TakJim committed:
remove unnecessary check

TakJim committed:
fix recon-all command syntax for normalization and comment typo

TakJim committed:
Fix for whitespace-safe syntax

TakJim committed:
Fix for whitespace-safe syntax

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+	fi
+	
+	log_Msg "...fourth recon-all steps with edited filled.mgz"
+	${recon_all_cmd} ${FILL} -tessellate -smooth1 -inflate1 -qsphere -fix -white -smooth2 -inflate2 -curvHK -sphere -surfreg -avgcurvtifpath $GCAdir -avgcurvtif $AvgCurvTif -jacobian_white -cortparc  ${extra_reconall_args} ${ExpertOpts}
```
Fixed it in 0a05d14e6ff41907488dbda3e77df7151e71fabf.

TakJim committed:
Fix the array logic for T2 argument

TakJim committed:
0a05d14e6ff41907488dbda3e77df7151e71fabf

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+		cmd=("$PipelineScripts"/RescaleVolumeAndSurface.sh "$SubjectDIR" "$SubjectID" "$SubjectDIR"/xfms/real2fs.world.mat "$T1wImage" "$T2wImage" "$T2Type" "$ScaleSuffix")
+		echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; ${cmd[@]} |& tee -a $LF
+		log_Msg "...fifth recon-all steps for hires white and pial using conf2hires"
+		${recon_all_cmd} -cortribbon ${recon_all_pial} ${ExpertOpts} ${conf2hiresflag}
```
Fixed it in 0a05d14e6ff41907488dbda3e77df7151e71fabf.

glasserm replied:
ExpertOpts removed, additional discussion needed?

TakJim committed:
Fix conditional check

TakJim committed:
remove unnecessary variables

TakJim committed:
Fixed conditional check

TakJim committed:
Fixed conditional check

TakJim committed:
Fixed conditional check

TakJim committed:
Fixed conditional check

TakJim committed:
Fixed conditional check

glasserm commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1137 @@
+	echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; ${cmd[@]} |& tee -a $LF
+
+	log_Msg "...third recon-all steps for segmentation with GCA"
+	${recon_all_cmd} -calabel -gca-dir $GCAdir -gca $GCA -normalization2 -maskbfs ${extra_reconall_args} ${ExpertOpts}
```
Again ExpertOpts removed?

TakJim committed:
fix default species to correct macaque species

TakJim committed:
Add RunLocal option and SGE submission

TakJim committed:
fix flag names to compatible with PiplineNHP.sh

glasserm commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1124 @@
+		recon_all_T2input=()
+		recon_all_pial=""
+		T2Type="NONE"		
+fi
```
Move this down...

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1124 @@
+	    recon_all_initrun+=(-i "$(remove_ext "$T1wImage")_scaled.nii.gz"
+	    -emregmask "$(remove_ext "$T1wImageBrain")_scaled.nii.gz")
+	fi
+	recon_all_initrun+=(${recon_all_T2input[@]+"${recon_all_T2input[@]}"}))
```
Stray extra `)`, my fault.
```suggestion
	recon_all_initrun+=(${recon_all_T2input[@]+"${recon_all_T2input[@]}"})
```

coalsont commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,174 @@
+for Subject in `echo $Subjlist | sed -e 's/@/ /g'` ; do
+
+    #Input Variables
+    Subject="$Subject" #FreeSurfer Subject ID Name
```
This line does nothing.
```suggestion
```

glasserm commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,174 @@
+  
+    "${queuing_command[@]}" ${HCPPIPEDIR}/FreeSurfer/FreeSurferPipelineNHP.sh \
+        --subject="$Subject" \
+        --subjectDIR="$SubjectDIR" \
```
It will save us from having to change it later.

TakJim replied:
Changed from --subject-dir to --session-dir to match Human Pipeline.
cd396fec0c7b5e09677bfb1e18d779103e39d095

TakJim commented on Examples/Scripts/FreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,174 @@
+        --flair="$isFLAIR" \
+        --species="$SPECIES" \
+        --runmode="$RunMode" \
+        --runlocal="$RunLocal"
```
Fixed it in https://github.com/Washington-University/HCPpipelines/commit/d5e73225d4b55d055088994092784adee8a655c2.

TakJim committed:
Change to --session-dir to match the Human pipeline

TakJim committed:
Move the recon_all_T2input configuration to just before it is used

TakJim committed:
remove extra )

TakJim committed:
remove unnecessary status check

TakJim committed:
remove unnecessary status check

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1124 @@
+    return_code=$?
+    if [ "${return_code}" != "0" ]; then
+        log_Err_Abort "mri_convert command failed with return code: ${return_code}"
+    fi
```
Fixed seven places in total.
8bfaa53e603bf69a9ee499f2bb31ed95ea3a15c7

TakJim committed:
remove unnecessary parameter configuration

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1086 @@
+
+		pushd ${mridir}
+
+		if [ "${flair}" = "TRUE" ] ; then
```
```suggestion
		if ((flair)) ; then
```

TakJim committed:
add capitalized flag

TakJim committed:
delete recon_all_T2input to simplify recon_all_initrun

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1086 @@
+            T2Type="NONE"		
+    fi
+
+	recon_all_initrun+=(${recon_all_T2input[@]+"${recon_all_T2input[@]}"})
```
Fixed it in d6d72fb301089d731de29b0389c239b912a7a567.

TakJim committed:
simplify $flair conditional branch

glasserm commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1083 @@
+fi
+if [[ ! -z "$GCSdir" && ! -z "$GCS" ]] ; then
+	extra_reconall_args+=(-gcs-dir $GCSdir -gcs $GCS)
+fi
```
```suggestion
if [[ -n "${GCSdir:-}" && -n "${GCS:-}" ]] ; then
    extra_reconall_args+=(-gcs-dir "$GCSdir" -gcs "$GCS")
fi

glasserm commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1083 @@
+		## This section scales them so that FreeSurfer 6 can work properly in scaled space. The data will be
+		## rescaled to the original space by a script, RescaleVolumeAndSurface.sh, after FS was finished - TH 2017-2023 
+		log_Msg "Scale T1w brain volume"		
+		${HCPPIPEDIR}/global/scripts/ScaleVolume.sh "${zero_threshold_T1wImage}" "$ScaleFactor" $(remove_ext ${T1wImage})_scaled "$SubjectDIR"/xfms/real2fs.world.mat
```
This code needs to be committed to the repository.

TakJim committed:
Add new file to change the size of the volume for NHP.

coalsont commented on global/scripts/ScaleVolume.sh:
```diff
@@ -0,0 +1,78 @@
+tmpmat=tmp_$$
+
+source $HCPPIPEDIR/global/scripts/log.shlib  # Logging related functions
+source $HCPPIPEDIR/global/scripts/opts.shlib # Command line option functions
```
I don't think this script uses anything in the old `opts.shlib`, so this can be removed.
```suggestion
```

coalsont commented on global/scripts/ScaleVolume.sh:
```diff
@@ -0,0 +1,78 @@
+	outmat=$4
+fi
+
+interp="${interp:-CUBIC}" # The script is irrelevant to resampling method.
```
If we want to allow different interpolation methods, please don't rely on `export` to get the value into the script, make it an optional argument.

TakJim committed:
remove unnecessary setting

TakJim committed:
simplify $conf2hires conditional branch

TakJim committed:
Handle unset GCS variables safely

TakJim committed:
Fix typo

TakJim committed:
remove unncessary shlib

TakJim committed:
Fix sourcing of log.shlib and debug.shlib

TakJim committed:
Make affine scaling robust when sform is NONE

TakJim committed:
Hardcode CUBIC interpolation

TakJim committed:
fix to use bash syntax

TakJim committed:
Hardcode CUBIC interpolation

TakJim committed:
Fix short-circuit evaluation

TakJim committed:
Handle unset optional argument safely

TakJim committed:
Handling unset options for conf2hires

TakJim committed:
Fix grep and tail usage

TakJim committed:
Fix scaling voxel size

TakJim committed:
fix newsform

TakJim committed:
Fix indentation and remove -lt 11 conditional branch

TakJim commented on global/scripts/ScaleVolume.sh:
```diff
@@ -0,0 +1,77 @@
+   newsform+=("$(echo "${sform[$i]}" | awk '{printf "%.8f\n",$1*'$ScaleFactor'}')")
+   if [ $i -lt 11 ] ; then     sumsform=$(echo "${sform[$i]}+${sumsform}" | bc -l)
+   fi
+done
```
Fixed it in fc596f5437af49bfda5ad09de566513a4d395afd.

takuya-hayashi replied:
I don't recall exactly why it was, perhaps, I guess it was typo of -le 11 or -lt 12?

coalsont commented on global/scripts/ScaleVolume.sh:
```diff
@@ -0,0 +1,76 @@
+# ----------------------------------------------------------------------
+log_Msg " reading sform from input"
+# ----------------------------------------------------------------------
+read -a sform <<<"$(wb_command -nifti-information -print-header "$T1wImage".nii.gz  | grep --text -A 3 "effective sform" | tail -n 3 | awk '{printf "%.8f\t%.8f\t%.8f\t%.8f\n",$1,$2,$3,$4}')"
```
deal with multiline sform issue (`tr` to replace newlines, instead of `-d ''` and `|| true`):
```suggestion
read -a sform <<<"$(wb_command -nifti-information -print-header "$T1wImage".nii.gz  | grep --text -A 3 "effective sform" | tail -n 3 | awk '{printf "%.8f\t%.8f\t%.8f\t%.8f\n",$1,$2,$3,$4}' | tr '\n' ' ')"
```

TakJim replied:
Fixed in 63b08d14243aed7c480eb04e36fada0eb505fe37.

glasserm commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1067 @@
+fi
+
+recon_all_cmd=(recon-all.v6.hiresNHP -subjid "$SubjectID" -sd "$SubjectDIR")
+extra_reconall_args+=(-openmp "$num_cores")
```
```suggestion

```

glasserm replied:
Resolved in [7ae8c17](https://github.com/Washington-University/HCPpipelines/pull/364/commits/7ae8c174925631df8b083ee7878c84863005103e)

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1067 @@
+# to have the "final say" over a couple settings within recon-all
+if ((conf2hires)); then
+	extra_reconall_args+=(-conf2hires)
+fi
```
Restore separate treatment of conf2hires.
```suggestion
#-conf2hires is more of a step than a setting, can't just include it in all calls
conf2hiresflag=""
if ((conf2hires)); then
	conf2hiresflag="-conf2hires"
fi
```

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1067 @@
+		echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; "${cmd[@]}" |& tee -a $LF
+		log_Msg "...fifth recon-all steps for hires white and pial using conf2hires"
+		"${recon_all_cmd[@]}" -cortribbon ${recon_all_pial} "${extra_reconall_args[@]}"
+	fi
```
...and use the separated conf2hires flag:
```suggestion
	if [ $(echo "$ScaleFactor < 6" | bc) = 1 ] ; then  # brain is larger than the rat
		log_Msg "...fifth recon-all steps for hires white and pial using conf2hires"
		#don't quote $recon_all_pial or $conf2hiresflag, we don't want to pass an empty string
		"${recon_all_cmd[@]}" -cortribbon ${recon_all_pial} "${extra_reconall_args[@]}" ${conf2hiresflag}
		log_Msg "...rescale volume and surface to native space"
		cmd=("$PipelineScripts"/RescaleVolumeAndSurface.sh "$SubjectDIR" "$SubjectID" "$SubjectDIR"/xfms/real2fs.world.mat "$T1wImage" "$T2wImage" "$T2Type" "$ScaleSuffix")
		echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; "${cmd[@]}" |& tee -a $LF
	else
		log_Msg "...rescale volume and surface to native space"
		cmd=("$PipelineScripts"/RescaleVolumeAndSurface.sh "$SubjectDIR" "$SubjectID" "$SubjectDIR"/xfms/real2fs.world.mat "$T1wImage" "$T2wImage" "$T2Type" "$ScaleSuffix")
		echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; "${cmd[@]}" |& tee -a $LF
		log_Msg "...fifth recon-all steps for hires white and pial using conf2hires"
		"${recon_all_cmd[@]}" -cortribbon ${recon_all_pial} "${extra_reconall_args[@]}" ${conf2hiresflag}
	fi
```

TakJim committed:
Fix multiline sform parsing in ScaleVolume.sh

TakJim committed:
Remove redundant OpenMP argument from recon-all command

TakJim committed:
Restore separate handling of -conf2hires

TakJim committed:
Add conf2hiresflag to recon_all_cmd

TakJim committed:
Add IntensityCor.sh

TakJim committed:
Add NHP-specific files

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1068 @@
+	## or aseg.presurf.edit.mgz was found, the script uses each as wm.mgz, brainmask.mnz, brain.finalsurfs.mgz and aseg.presurf.mgz respectively.
+	## ${PipelineScripts}/IntensityNormalize.sh may be useful for normalizing intensity of white matter or grey matter to create
+	## brain.finalsurfs.edit.mgz,
+	cmd=("$PipelineScripts"/SubcortSegment.sh "$SubjectDIR" "$SubjectID" "$T1wImage" "$TemplateWMSkeleton" "$SubjectDIR"/xfms/real2fs.world.mat "$mri_segment_args")
```
fix undefined/multiple arguments in currently-unused `mri_segment_args`:
```suggestion
	cmd=("$PipelineScripts"/SubcortSegment.sh "$SubjectDIR" "$SubjectID" "$T1wImage" "$TemplateWMSkeleton" "$SubjectDIR"/xfms/real2fs.world.mat ${mri_segment_args:-${mri_segment_args}})
```

TakJim committed:
Fix mri_segment_args handling in SubcortSegment call

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1083 @@
+	echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; ${cmd[@]} |& tee -a $LF
+
+	## This section replaces 'FS -nuintensirycor' for NHP - TH 2017-2023
+	cmd=("$PipelineScripts"/IntensityCor.sh "$SubjectDIR"/"$SubjectID"/mri/orig.mgz "$SubjectDIR"/"$SubjectID"/mri/brainmask.conf.mgz "$SubjectDIR"/"$SubjectID"/mri/nu.mgz -t1 -m "$IntensityCor","$BiasFieldFastSmoothingSigma" "$StrongBias_args")
```
Looks like another `SetUpFSNHP.sh` variable that isn't set for all species.
```suggestion
	cmd=("$PipelineScripts"/IntensityCor.sh "$SubjectDIR"/"$SubjectID"/mri/orig.mgz "$SubjectDIR"/"$SubjectID"/mri/brainmask.conf.mgz "$SubjectDIR"/"$SubjectID"/mri/nu.mgz -t1 -m "$IntensityCor","$BiasFieldFastSmoothingSigma" ${StrongBias_args:-})
```

TakJim committed:
Fix handling of unbounded StrongBias_args

TakJim committed:
Make command execution whitespace-safe

takuya-hayashi commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,274 @@
+  else                                                        # control T2-FLAIR pial
+    NSigmaAbove="3"                                           # T2w/FLAIR version, default=3 
+    NSigmaBelow="4"                                           # T2w/FLAIR version, default=3  
+  fi    
```
@TakJim please change the settings between T2 pial and T2 FLAIR 
```suggestion
  if ((isFLAIR)) ; then                                              # control T2-FLAIR pial
    NSigmaAbove="3"                                           # T2w/FLAIR version, default=3 
    NSigmaBelow="4"                                           # T2w/FLAIR version, default=3
  else                                                                     # control T2 pial
    NSigmaAbove="0"                                           # 2: FS6 default
    NSigmaBelow="3"                                           # 3: FS6 default
  fi

glasserm replied:
Currently isFLAIR is a true/false variable.  I would prefer there were a separate variable that states the data will be divided.

takuya-hayashi replied:
``` suggestion
  if ((isFLAIR)) ; then                                       # control T2w-FLAIR pial tuning
    if ((DivideT1wForPial)); then                             # control with T1w/FLAIR
      NSigmaAbove="3"                                         # 3: highres FLAIR (Autio et al.)                                        
      NSigmaBelow="4"                                         # 4: highres FLAIR (Autio et al.)                                         
    else                                                      # control with FLAIR
      NSigmaAbove="3"                                         # 3: FS6 default
      NSigmaBelow="3"                                         # 3: FS6 default
    fi
  else                                                        # control with T2w
    NSigmaAbove="0"                                           # 2: FS6 default
    NSigmaBelow="3"                                           # 3: FS6 default
  fi

takuya-hayashi reviewed (commented):
@TakJim please exchange the value settings between T2 pial and T2 FLAIR

glasserm commented on FreeSurfer/custom/recon-all.v6.hiresNHP:
```diff
@@ -0,0 +1,9229 @@
+  if($RunIt) $fs_time $cmd |& tee -a $LF
+  if($status) goto error_exit;
+  echo $cmd > $touchdir/conf2hires
+endif
```
Call with T1w/FLAIR or just FLAIR argument.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1068 @@
+
+	## This section replaces 'FS -nuintensirycor' for NHP - TH 2017-2023
+	cmd=("$PipelineScripts"/IntensityCor.sh "$SubjectDIR"/"$SubjectID"/mri/orig.mgz "$SubjectDIR"/"$SubjectID"/mri/brainmask.conf.mgz "$SubjectDIR"/"$SubjectID"/mri/nu.mgz -t1 -m "$IntensityCor","$BiasFieldFastSmoothingSigma" ${StrongBias_args:-})
+	echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; ${cmd[@]} |& tee -a $LF
```
Please apply this change to the other ~20 identical lines below in the script:
```suggestion
	echo -e "$(date)\n#===============================\n${cmd[*]}\n" |& tee -a "$LF"; "${cmd[@]}" |& tee -a "$LF"

	cmd=(mri_convert "$SubjectDIR"/"$SubjectID"/mri/brainmask.orig.nii.gz "$SubjectDIR"/"$SubjectID"/mri/brainmask.conf.mgz --conform)
	echo -e "$(date)\n#===============================\n${cmd[*]}\n" |& tee -a "$LF"; "${cmd[@]}" |& tee -a "$LF"

	## This section replaces 'FS -nuintensirycor' for NHP - TH 2017-2023
	cmd=("$PipelineScripts"/IntensityCor.sh "$SubjectDIR"/"$SubjectID"/mri/orig.mgz "$SubjectDIR"/"$SubjectID"/mri/brainmask.conf.mgz "$SubjectDIR"/"$SubjectID"/mri/nu.mgz -t1 -m "$IntensityCor","$BiasFieldFastSmoothingSigma" ${StrongBias_args:-})
	echo -e "$(date)\n#===============================\n${cmd[*]}\n" |& tee -a "$LF"; "${cmd[@]}" |& tee -a "$LF"
```

TakJim committed:
Fix swapped isFLAIR/non-FLAIR NSigma values in SetUpFSNHP.sh

TakJim commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,274 @@
+
+  # FS BBR in DiffusionPreprocPipeline
+  FSBBRDIFF=TRUE
+  DiffWMProjAbs="2"
```
Rebased onto master (HEAD: 8dd5aa6).

TakJim committed:
Add T1w/FLAIR mode

TakJim committed:
Use tempfiles.shlib for temporary matrix file in ScaleVolume.sh

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1068 @@
+		echo -e "$(date)\n#===============================\n${cmd[@]}\n" |& tee -a $LF; "${cmd[@]}" |& tee -a $LF
+		log_Msg "...fifth recon-all steps for hires white and pial using conf2hires"
+		"${recon_all_cmd[@]}" -cortribbon ${recon_all_pial} "${extra_reconall_args[@]}" ${conf2hiresflag}
+	fi
```
Fixed it in 0888b46f.

TakJim committed:
Make echo/cmd logging whitespace-safe and quote LF variable

TakJim committed:
Move FSBBRDIFF/DiffWMProjAbs from SetUpFSNHP.sh to SetUpSPECIES.sh

coalsont commented on FreeSurfer/scripts/SubcortSegment.sh:
```diff
@@ -0,0 +1,213 @@
+#!/bin/bash
```
When does this script get used?

glasserm replied:
We will keep this.

coalsont commented on FreeSurfer/scripts/RevertPreFreeSurferResampling.sh:
```diff
@@ -0,0 +1,70 @@
+#! /bin/bash
```
Also, when does this script get used?

TakJim committed:
Fix simple review comments on PR #364

- ScaleVolume.sh: Remove unnecessary exit 0
- SubcortSegment.sh: Use debug.shlib instead of opts.shlib, simplify log_Msg
- FreeSurferPipelineNHP.sh: Remove unnecessary local t1wflair declaration
- FreeSurferPipeline.sh: Revert accidental t1wflair addition to human pipeline

TakJim commented on global/scripts/ScaleVolume.sh:
```diff
@@ -0,0 +1,75 @@
+# ----------------------------------------------------------------------
+log_Msg "End: ScaleVolume.sh"
+# ----------------------------------------------------------------------
+exit 0
```
Fixed in 7352a708.

TakJim commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1076 @@
+    local mri_vol2vol_cmd
+    local return_code
+    local t2_or_flair
+    local t1wflair
```
Fixed in 7352a708.

TakJim commented on FreeSurfer/FreeSurferPipeline.sh:
```diff
@@ -622,6 +622,8 @@ if [[ "${T2wImage}" != "" ]]; then
     if ((flair)); then
         t2_or_flair="FLAIR"
+    else if ((t1wflair)); then
+        t2_or_flair="T1wFLAIR"
```
Reverted in 7352a708.

TakJim committed:
Rename T1wFLAIR to T1wDivFLAIR throughout for clarity

Rename variables, CLI flags, and string values from T1wFLAIR/t1wflair
to T1wDivFLAIR/t1wdivflair to be less confusing, as suggested in PR review.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1075 @@
+	## or aseg.presurf.edit.mgz was found, the script uses each as wm.mgz, brainmask.mnz, brain.finalsurfs.mgz and aseg.presurf.mgz respectively.
+	## ${PipelineScripts}/IntensityNormalize.sh may be useful for normalizing intensity of white matter or grey matter to create
+	## brain.finalsurfs.edit.mgz,
+	cmd=("$PipelineScripts"/SubcortSegment.sh "$SubjectDIR" "$SubjectID" "$T1wImage" "$TemplateWMSkeleton" "$SubjectDIR"/xfms/real2fs.world.mat ${mri_segment_args:-${mri_segment_args}})
```
My fault for using syntax like for empty-array handling.  I didn't realize it still did undefined checks in the replacement context.

```suggestion
	cmd=("$PipelineScripts"/SubcortSegment.sh "$SubjectDIR" "$SubjectID" "$T1wImage" "$TemplateWMSkeleton" "$SubjectDIR"/xfms/real2fs.world.mat ${mri_segment_args:-})
```

TakJim replied:
Fixed in 0bf01a5b.

TakJim committed:
Fix review comments on PR #364

- Pass $t1wdivflair argument to SetUpFSNHP.sh
- Fix mri_segment_args empty-array handling syntax

TakJim committed:
Remove --no-T1wDivFLAIR flag that overrides FLAIR in conf2hires call

The --no-T1wDivFLAIR option was overriding the preceding --FLAIR flag
because recon-all only accepts one mode and the last one takes
precedence. Since the default already implies no T1wDivFLAIR, this
flag is unnecessary and harmful when FLAIR is enabled.

TakJim committed:
Fix SetUpFSNHP.sh: move T1w/FLAIR pial params from Chimp to Macaque

Resolves PR #364 comments from glasserm (2026-03-16):
- Remove isT1wDivFLAIR branch from Chimp section (was misplaced)
- Add isT1wDivFLAIR branch to Macaque section with correct values
  (NSigmaAbove=3, NSigmaBelow=4, per Autio et al.)
- Exchange isFLAIR values in Macaque: was using T1w/FLAIR values (3,4)
  for FLAIR mode, now correctly uses (0,3) for FLAIR and (3,4) for
  T1w/FLAIR mode

This was causing pial surface to collapse onto white surface when
running T1w/FLAIR mode on macaque data (wrong NSigma parameters).

TakJim committed:
Revert isFLAIR values to original (3,4) in Macaque section

Matt only said Move to Macaque - no instruction to change the values.
Restore original isFLAIR NSigma values (3,4) in Macaque section.

TakJim committed:
SetUpFSNHP.sh: set isFLAIR NSigma to default (3,3) and add isT1wDivFLAIR branch for Marmoset

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1075 @@
+		log_Msg "revert PreFreeSurfer resampling"	
+		${HCPPIPEDIR_FS}/RevertPreFreeSurferResampling.sh $(dirname $(dirname "$SubjectDIR")) "$SubjectID"
+		imrm ${SubjectDIR}/xfms/${OutputOrigT2wToT1w}
+	fi
```
"revert PreFreeSurfer resampling" suggests that the effect is to restore things to how they were *before* PreFS, which according to these comments is not the actual effect.  Consider something like `UndoNHPPostFreeSurferResampling.sh` or `RedoNHPPreFreeSurferResampling.sh` (or "restore") instead, for clarity.

TakJim replied:
Renamed to RestorePreFreeSurferResamplingNHP.sh and updated all references in 92d8eddb.

TakJim committed:
Rename IntensityCor.sh to IntensityCorNHP.sh

NHP-specific sub-scripts should have "NHP" in their names.

TakJim committed:
Fix header comment in IntensityCorNHP.sh

TakJim committed:
Address review comments: remove IsRunning cleanup, rename script, fix conf2hiresNHP conditionals

- Remove IsRunning.lh+rh cleanup code that defeats FreeSurfer internal locking (coalsont/glasserm)
- Rename RevertPreFreeSurferResampling.sh to RestorePreFreeSurferResamplingNHP.sh (glasserm)
- Fix conf2hiresNHP: separate DoFLAIR and DoT1wDivFLAIR into independent if blocks (glasserm)

TakJim committed:
Pass ScaleFactor from SetUpSPECIES.sh instead of duplicating in SetUpFSNHP.sh

- BatchNHP: source SetUpSPECIES.sh to get BrainScaleFactor, add StrucRes
- FreeSurferPipelineNHP: add --scale-factor mandatory option, pass to SetUpFSNHP
- SetUpFSNHP: receive ScaleFactor as 4th argument, remove per-species hardcoded values

Eliminates duplicate ScaleFactor definitions that could diverge from
BrainScaleFactor in SetUpSPECIES.sh (coalsont review comment).

coalsont commented on FreeSurfer/scripts/RescaleVolumeAndSurface.sh:
```diff
@@ -0,0 +1,229 @@
+#! /bin/bash
+# RescaleVolumeAndSurface.sh
+
+#set -eu
+
+Usage_exit (){
+echo "RescaleVolumeAndSurface.sh <SubjectDIR> <SubjectID> <ScaleVolumeMatrix (world.mat)> <T1wImage> <T2wImage> <FLAIR|T2|NONE> [ScaleSuffix]"
+exit 0;
+}
```
Is this expected to be useful outside of NHP?

Is this undoing the scaling done by `ScaleVolume.sh`?  Might be good for the help to provide a little context.

TakJim replied:
Fixed it in 0036c54a. (The script file is already named RescaleVolumeAndSurfaceNHP.sh.)

coalsont commented on FreeSurfer/custom/SetUpFSNHP.sh:
```diff
@@ -0,0 +1,257 @@
+SPECIES="$1"
+isFLAIR="$2"    #NOTE: must be "0" or "1"
+isT1wDivFLAIR="$3" #NOTE: must be "0" or "1"
+ScaleFactor="$4" # must match BrainScaleFactor in SetUpSPECIES.sh
```
To be clear, the better solution is probably to compute the sigmas from `ScaleFactor`, but we'd need to check each species individually to see whether or not the "20 * ScaleFactor" and similar comments are actually accurate.

glasserm replied:
Receive bias field sigmas are related to the coil design, not necessarily the brain (though there may be a correlation because certain coils are used for certain brains).

coalsont replied:
Part of the conflict here is that `ScaleFactor` is currently provided as an argument to `FreeSurferNHP.sh`, but appears to have originally been designed to be hardcoded in `SetUpFSNHP.sh`.  Is it important to give users control over `ScaleFactor`?  If not, we can remove the `--scale-factor` mandatory parameter, hardcode it back into `SetUpFSNHP.sh`, and this discussion becomes moot.

takuya-hayashi replied:
Let's define it in SetUpSPECIES.sh.
1) It takes precedence over SetUpFSNHP.sh.
2) BrainScaleFactor should usually be the same as ScaleFactor in FS.
3) BrainScaleFactor is not currently used in PreFS; however, it would be beneficial to include a function in the future to improve registration with FLIRT and FNIRT in the scaled space and rescale to the original space. I'm now preparing 'flirt_scale' and 'fnirt_scale'.

coalsont replied:
"usually" suggests these parameters are intended for independent purposes.  My understanding is that `ScaleFactor` is for tricking freesurfer into handling NHP images, because it expects a human-sized brain (and after rescaling and converting freesurfer files to what we need, generally won't be needed by later pipelines, with the possible loose end of `bbregister`).  What is `BrainScaleFactor` intended to be used for?

mmilch01 replied:
Make into a single variable named BrainScaleFactor

coalsont replied:
`ScaleFactor` appears to have been removed from SetUpFSNHP already, so it may currently only be a variable name inside the FS pipeline, which is fine.  It sounds like we don't currently need to implement user control over the parameters that were hardcoded in tandem with the previously-hardcoded values of ScaleFactor (though we might want it in the future).  So, this may be resolved.

coalsont commented on FreeSurfer/scripts/Conf2GCAReg_FNIRTbased.sh:
```diff
@@ -0,0 +1,104 @@
+
+source $HCPPIPEDIR/global/scripts/log.shlib  # Logging related functions
+source $HCPPIPEDIR/global/scripts/opts.shlib # Command line option functions
+log_SetToolName "$(basename $0)"
```
Doesn't look like the old `opts.shlib` is used.  Simplify the log stuff, too.

```suggestion
source $HCPPIPEDIR/global/scripts/log.shlib "$@" # Logging related functions
```

coalsont commented on FreeSurfer/scripts/RescaleVolumeAndSurface.sh:
```diff
@@ -0,0 +1,229 @@
+t2_or_flair="$6"
+ScaleSuffix="${7:-_scaled}"
+
+log_SetToolName "RescaleVolumeAndSurface.sh"
```
Redundant, `log.shlib` already sets it from `$0` regardless.

```suggestion
```

takuya-hayashi commented on FreeSurfer/scripts/IntensityNormalize.sh:
```diff
@@ -0,0 +1,181 @@
+
+# Normalize intensity of cortical grey or white matter of input volumes
+# Takuya Hayashi, RIKEN Brain Connectomics Imaging Laboratory, 2024
+# set -eux
```
Probably no.

TakJim committed:
Apply review suggestions: simplify source/log lines, fix descriptions, remove unused code

TakJim committed:
Rename Revert→Restore in RestorePreFreeSurferResamplingNHP.sh and fix log messages

TakJim committed:
Rename SubcortSegment, IntensityNormalize, ScaleVolume with NHP suffix

TakJim commented on global/scripts/ScaleVolumeNHP.sh:
```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+# ScaleVolume.sh
+# Scale volume for NHP FreeSurfer
```
Renamed in c1a6ff42.

TakJim commented on FreeSurfer/scripts/SubcortSegmentNHP.sh:
```diff
@@ -0,0 +1,211 @@
+#!/bin/bash
+# SubcortSegment.sh
+# The script edits subcortical segment (aseg) and white matter segment (wm) for NHP brain using FS6. The wm is deweighted
```
Renamed in c1a6ff42.

TakJim committed:
Rename RescaleVolumeAndSurface, Conf2GCAReg_FNIRTbased with NHP suffix

TakJim committed:
Revert MMode=T2 back to MMode=FLAIR in conf2hiresNHP

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1074 @@
+
+opts_AddOptional '--runmode' 'RunMode' 'Default|FSinit|FSbrainseg|FSsurfinit|FShires|FSFinish' "specify from which step to resume the processing instead of starting from the beginning. Value must be one of: Default, FSinit, FSbrainseg, FSsurfinit, FShires, FSFinish (default: Default)" "Default"
+
+opts_AddMandatory '--scale-factor' 'ScaleFactor' 'number' "Brain scale factor for NHP processing (must match BrainScaleFactor in SetUpSPECIES.sh).  Controls volume rescaling for FreeSurfer and derived parameters in SetUpFSNHP.sh."
```
Sorry, I forgot that FreeSurfer having a separate NHP pipeline script is the exception, not the rule.  Also, I finally went and checked PreFS, and there isn't actually a scaling factor there.

So really, this comes back to the discussion that SetUpFSNHP.sh sets hardcoded sigmas based on an assumed scale factor.

mmilch01 replied:
The same solution - make into a single variable named BrainScaleFactor

coalsont replied:
To be clear, because this is a parameter to the script, it is acceptable for the variable name inside of `FreeSurferPipelineNHP.sh` to be different than in a SetUp file, so it can stay `ScaleFactor` in this script.

TakJim committed:
Add description to RescaleVolumeAndSurfaceNHP usage

TakJim committed:
ScaleVolumeNHP: simplify sform read using existing float formatting

TakJim committed:
RescaleVolumeAndSurfaceNHP: drop redundant END log_Msg and exit 0

TakJim commented on global/scripts/ScaleVolumeNHP.sh:
```diff
@@ -0,0 +1,72 @@
+# ----------------------------------------------------------------------
+log_Msg " reading sform from input"
+# ----------------------------------------------------------------------
+read -d '' -a sform <<<"$(wb_command -nifti-information -print-header "$T1wImage".nii.gz  | grep --text -A 3 "effective sform" | tail -n 3 | awk '{printf "%.8f\t%.8f\t%.8f\t%.8f\n",$1,$2,$3,$4}')" || true
```
Fixed it in 8afffe79.

TakJim commented on FreeSurfer/scripts/RescaleVolumeAndSurfaceNHP.sh:
```diff
@@ -0,0 +1,228 @@
+log_Msg "END: RescaleVolumeAndSurface"
+
+exit 0
+
+
```
Fixed it in debd2dca.

TakJim committed:
IntensityCorNHP: migrate to newopts.shlib

Replace positional arguments and getopts with newopts.shlib named
options (--input, --brainmask, --output, --type, --wm-mask, --method,
--smoothing, --strongbias). Update call site in FreeSurferPipelineNHP
and rename StrongBias_args to StrongBias in SetUpFSNHP.

TakJim commented on FreeSurfer/scripts/IntensityCorNHP.sh:
```diff
@@ -0,0 +1,160 @@
+fi
+
+strongbiasflag=""
+while getopts m:s opt; do
```
Migrated to `newopts.shlib` in b5c79f00.

Replaced positional arguments and `getopts` with named options (`--input`, `--brainmask`, `--output`, `--type`, `--wm-mask`, `--method`, `--smoothing`, `--strongbias`). Also renamed `StrongBias_args` to `StrongBias` (TRUE/FALSE) in `SetUpFSNHP.sh` and updated the call site in `FreeSurferPipelineNHP.sh`.

coalsont replied:
Looks good, thanks.

coalsont commented on FreeSurfer/scripts/IntensityCorNHP.sh:
```diff
@@ -0,0 +1,159 @@
+    log_Err_Abort "unknown --type '$ImageType', must be T1 or T2"
+fi
+
+type="$ImageType"
```
It would be preferable not to have two variables doing the same thing.

mmilch01 replied:
Use 'ImageType' instead of 'type' throughout.

TakJim replied:
Fixed in 142f28c6. Removed the `type="$ImageType"` alias and use `ImageType` throughout the script.

coalsont commented on FreeSurfer/FreeSurferPipelineNHP.sh:
```diff
@@ -0,0 +1,1074 @@
+
+opts_AddOptional '--runmode' 'RunMode' 'Default|FSinit|FSbrainseg|FSsurfinit|FShires|FSFinish' "specify from which step to resume the processing instead of starting from the beginning. Value must be one of: Default, FSinit, FSbrainseg, FSsurfinit, FShires, FSFinish (default: Default)" "Default"
+
+opts_AddMandatory '--scale-factor' 'ScaleFactor' 'number' "Brain scale factor for NHP processing (must match BrainScaleFactor used during PreFreeSurferNHP.sh).  Controls volume rescaling for FreeSurfer and derived parameters in SetUpFSNHP.sh."
```
Make the usage wording less strict, since there isn't something from a prior pipeline that this would need to match.

```suggestion
opts_AddMandatory '--scale-factor' 'ScaleFactor' 'number' "Brain scale factor for NHP processing (recommended value is BrainScaleFactor in SetUpSPECIES.sh).  Controls volume rescaling for FreeSurfer and some derived parameters in SetUpFSNHP.sh."
```

TakJim replied:
Applied your suggestion in 9ae9e978.

TakJim committed:
IntensityCorNHP: drop type alias, use ImageType throughout

TakJim committed:
FreeSurferPipelineNHP: soften --scale-factor usage wording

TakJim committed:
Merge branch 'master' into NHP_FreeSurferPipeline

coalsont commented:
The conflict resolution (after my edit to the fMRI parameters on master) appears to have deleted lines like these from `SetUpSPECIES.sh`:

```
    # FS BBR in DiffusionPreprocPipeline
    FSBBRDIFF=TRUE
    DiffWMProjAbs="1"
```
We should probably restore them, though maybe they belong in the diffusion PR.  For reference, here they are before the conflict resolution commit:

https://github.com/Washington-University/HCPpipelines/blob/9ae9e978b161171c936f1f9f54fce08f779b6987/Examples/Scripts/SetUpSPECIES.sh

coalsont commented on FreeSurfer/scripts/RescaleVolumeAndSurfaceNHP.sh:
```diff
@@ -0,0 +1,226 @@
+log_Msg "Rescaling volumes"
+# ----------------------------------------------------------------------
+# Rescaling voxel size (note that data is not virtually resampled). Thus resampling type (e.g. trilinear vs nearest) is irrelevant.
+if [[ ${freesurfer_primary_version} = 6 && $(echo "$ScaleFactor < 6" | bc) = 1 ]] ; then  # brain is larger than the rat
```
Is this freesurfer version check intended to distinguish 6.x from 5.3.0, or from 7+?  It seems unlikely that we would want to do the same thing on versions both older and newer than 6, so `-ge` or `-le` (or the more obvious `((freesurfer_primary_version <= 6)) && [[ ...` and similar) would make more sense.

glasserm replied:
@takuya-hayashi please address with @TakJim.

glasserm replied:
Tim to make code suggestion.

coalsont replied:
```suggestion
if ((freesurfer_primary_version >= 6)) && [[ $(echo "$ScaleFactor < 6" | bc) == 1 ]] ; then  # brain is larger than the rat
```

TakJim replied:
Fixed in f8713e7a — applied your suggestion verbatim on both L154 and L186.

coalsont commented on FreeSurfer/scripts/RescaleVolumeAndSurfaceNHP.sh:
```diff
@@ -0,0 +1,226 @@
+# ----------------------------------------------------------------------
+log_Msg "Rescaling or copying surface data"
+# ----------------------------------------------------------------------
+if [[ ${freesurfer_primary_version} = 6 && $(echo "$ScaleFactor < 6" | bc) = 1 ]] ; then
```
Same issue as above.

glasserm replied:
Tim to make code suggestion.

coalsont replied:
```suggestion
if ((freesurfer_primary_version >= 6)) && [[ $(echo "$ScaleFactor < 6" | bc) == 1 ]] ; then
```

TakJim replied:
Fixed in f8713e7a.

coalsont commented on FreeSurfer/scripts/SubcortSegmentNHP.sh:
```diff
@@ -0,0 +1,211 @@
+			fi
+		fi
+
+		mv brainmask.edit.mgz brainmask.mgz
```
Some of these `checkfile=` sections seem to be trying to predict whether the next command will fail (we generally instead rely on the command to print an understandable error message, exit with nonzero status, which is then picked up by `debug.shlib` (which is sourced in this script), so we don't usually have to do any manual checking per command).

Others, like this one, seem to be testing some file other than the one that is about to be overwritten, which seems like either a different purpose, or copy/paste errors.  Can you check the filenames and/or clarify the intent of these (via comments in the code if they are not simply checking for failure)?

glasserm replied:
@takuya-hayashi please address with @TakJim.

glasserm replied:
Let's remove these.

TakJim replied:
Removed all 8 `checkfile=` writability check blocks in `SubcortSegmentNHP.sh` in commit 9afbcd57. The blocks tried to predict whether the next `mv`/`cp` would fail by pre-checking write permission, but the actual command will fail with a useful filesystem error anyway. Takuya confirmed at 2026-04-23 MTG.

glasserm commented:
> The conflict resolution (after my edit to the fMRI parameters on master) appears to have deleted lines like these from `SetUpSPECIES.sh`:
> 
> ```
>     # FS BBR in DiffusionPreprocPipeline
>     FSBBRDIFF=TRUE
>     DiffWMProjAbs="1"
> ```
> 
> We should probably restore them, though maybe they belong in the diffusion PR. For reference, here they are before the conflict resolution commit:
> 
> https://github.com/Washington-University/HCPpipelines/blob/9ae9e978b161171c936f1f9f54fce08f779b6987/Examples/Scripts/SetUpSPECIES.sh

I agree these should be added to the diffusion PR.

mmilch01 commented on FreeSurfer/custom/conf2hiresNHP:
Please set execute permission on this file.

TakJim replied:
Fixed in 0dae829c — set executable permission via `git update-index --chmod=+x` to match the human `conf2hires` (which is 100755).

mmilch01 commented on FreeSurfer/custom/recon-all.v6.hiresNHP:
Please set execute permission on this file.

TakJim replied:
Fixed in 0dae829c — same fix as the conf2hiresNHP thread, set 100755 to match the human counterpart.

TakJim committed:
SubcortSegmentNHP.sh: remove checkfile= writability checks

Remove all 8 checkfile= blocks per PR #364 review (Matt
glasserm r3121341520, r3127122715, also Tim's request from
2026-04-23 NHP Pipelines MTG).

The blocks tried to predict whether the next mv/cp command
would fail by pre-checking write permission on the destination.
This is over-engineering: the next command will fail with a
useful filesystem error anyway, and predicting failure outside
the actual write target only catches a narrow case.

Takuya confirmed the removal at 2026-04-23 MTG.

TakJim committed:
RescaleVolumeAndSurfaceNHP.sh: use >=6 for FS version check

Per Tim suggestions r3121196941 and r3121199083, the strict
equality check ${freesurfer_primary_version} = 6 excluded FS 7
from the v6+ codepath, which was almost certainly unintended.
Switch to arithmetic comparison ((freesurfer_primary_version >= 6))
on both the volume rescaling block (L154) and the surface
rescaling block (L186).

Also align the inner ScaleFactor test from = 1 to == 1.
Behavior unchanged for FS 6 itself; FS 7 now follows the same
path as FS 6.

TakJim committed:
custom: set execute permission on conf2hiresNHP and recon-all.v6.hiresNHP

Per mmilch01 r3163433523 and r3163434561: the NHP variants
of conf2hires and recon-all.v6.hires were committed with
mode 100644 while their human counterparts are 100755. They
are sourced/run as scripts so they need +x.

mmilch01 commented on FreeSurfer/FreeSurferPipelineNHP.sh:
Please set execute permission on this script.

TakJim replied:
Set executable permission in 04e22245.

mmilch01 commented on FreeSurfer/custom/SetUpFSNHP.sh:
Please set execute permission on this file. Otherwise, 'which' operator on line 255 of FreeSurferPipelineNHP.sh will fail.

TakJim replied:
Set executable permission in ef9455e3.

TakJim committed:
FreeSurferPipelineNHP.sh: set executable permission

Per mmilch01 review on PR #364.

TakJim committed:
SetUpFSNHP.sh: set executable permission

Per mmilch01 review on PR #364 — needed so that 'which' on line 255 of
FreeSurferPipelineNHP.sh resolves it.
